qiskit/algorithms/gradients/base_estimator_gradient.py

```diff
@@ -0,0 +1,122 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Abstract Base class of Gradient for Estimator.
+"""
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from collections.abc import Sequence
+
+from qiskit.circuit import Parameter, QuantumCircuit
+from qiskit.exceptions import QiskitError
+from qiskit.opflow import PauliSumOp
+from qiskit.primitives import BaseEstimator
+from qiskit.quantum_info.operators.base_operator import BaseOperator
+
+from .estimator_gradient_result import EstimatorGradientResult
+
+
+class BaseEstimatorGradient(ABC):
+    """Base class for an EstimatorGradient to compute the gradients of the expectation value."""
+
+    def __init__(
+        self,
+        estimator: BaseEstimator,
+        **run_options,
+    ):
+        """
+        Args:
+            estimator: The estimator used to compute the gradients.
+        """
+        self._estimator: BaseEstimator = estimator
+        self._circuits: Sequence[QuantumCircuit] = []
+        self._circuit_ids: dict[int, int] = {}
```

Comment:
Do we need `_circuit_ids`? The mechanism for matching circuits by object id doesn't seem necessary for this layer, is it difficult to implement without this? Originally this is a mechanism to reduce the communication cost of RuntimeEstimator/Sampler.

Reply:
I think it can help you save some time, for example not constructing the gradient circuits multiple times [here](https://github.com/Qiskit/qiskit-terra/blob/f918a2a235361a16758babb945be2cbfad9b38a6/qiskit/algorithms/gradients/lin_comb_estimator_gradient.py#L72)

Reply:
For `dict[int, int]` it might be preferable, like the prior one uses Sequence, for this to be `Mapping[int, int]`.

Default values like `[]` and `{}` are mutable and since Python maintains a single global instance that it reuses for this default value any change to it internally will affect all instances. So for mutable defaults its more that you would default to None here and internally if None set it to [] or {} which will then be specific to the instance and even if changed will not affect other instances.

Reply:
To @ikkoham, as @ElePT mentioned, the gradients cache circuits because we need some preprocessing (e.g. preparing gradient circuits). By caching circuits, we can avoid doing the preprocessing task multiple times. 

@woodsp-ibm, thank you! I didn't know about it. I changed the codes.

Reply:
For example, if you want to cache a circuit, wouldn't it be straighter forward to have a cache of object_id => QuantumCircuit (Mapping[int, QuantumCircuit])?

Reply:
I thought the same last night for the fidelity class, and I agree that it's more straightforward ([here](https://github.com/ElePT/qiskit-terra/blob/fidelity-primitive/qiskit/algorithms/fidelities/base_fidelity.py#L40) is my implementation).

Reply:
That probably works too. I was thinking to access the gradient circuits and the base parameter values with `circuit_index`. (Since we can update circuits in the primitives, maybe we don't need this circuit index variable.)
  
              circuit_index = len(self._circuits)
                self._circuit_ids[id(circuit)] = circuit_index
                self._gradient_circuit_data_dict[
                    circuit_index
                ] = make_param_shift_gradient_circuit_data(circuit)
                self._base_parameter_values_dict[
                    circuit_index
                ] = make_param_shift_base_parameter_values(
                    self._gradient_circuit_data_dict[circuit_index]
                )

https://github.com/a-matsuo/qiskit-terra/blob/f19222dbf87ddf82b54d527cb31e26e8143c5cb5/qiskit/algorithms/gradients/param_shift_estimator_gradient.py#L75-L84